source/common/upstream/upstream_impl.cc

```diff
@@ -102,6 +133,16 @@ HostImpl::createConnection(Event::Dispatcher& dispatcher, const ClusterInfo& clu
   } else {
     cluster_options = options;
   }
+  if (cluster.features() & ClusterInfo::Features::USE_TCP_KEEPALIVE) {
```

Comment:
Instead of adding a big to features(), can you just condition on whether cluster.tcpKeepaliveSettings() is set or not, and turn it into an optional?

Reply:
I copied this pattern from the HTTP2 feature, I assumed there was maybe a performance reason for using the features flags rather than checking an optional field. If there's not, then I'll change it.

Reply:
Looking at this further, I'm wondering if I should actually refactor to instantiate the TcpKeepaliveOptionImpl (and also UpstreamSocketOption for freebind) at ClusterInfoImpl construct time and store them both in a Options vector on ClusterInfo.
That removes them both from Features.
That should also greatly simplify the code in createConnection, since there's no need to have different cases for them (it's just a case of merging options vectors together is necessary.)
This should be safe as neither Option implementation stores any state that differs between connections to the same cluster.

What do you think?

Reply:
My best guess for why this is the way it is: the features flags came befofe the current iteration of socket-option-setters.

I think your suggested refactoring sounds great, and should be a nice cleanup. @htuch do you know any more history on this, or any objections to this cleanup?

Reply:
I've made an attempt at the refactor (based off this branch) in this commit https://github.com/JonathanO/envoy/commit/bf4bf78b3739529f19d899e5dfabdca9ee1f6196 which I think looks a bit cleaner.

I must admit that I'm (re)learning modern C++ by trial and error, and I have no idea if the shared pointer usage in that commit is right or even necessary.
If it looks sane then I'll merge it into this PR.